src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java

```diff
@@ -0,0 +1,123 @@
+import uk.gov.hmcts.reform.sscs.service.HmcHearingApiService;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ProcessHmcMessageServiceV2 {
+
+    private final HmcHearingApiService hmcHearingApiService;
+    private final HearingUpdateService hearingUpdateService;
+    private final UpdateCcdCaseService updateCcdCaseService;
+    private final IdamService idamService;
+    private final ProcessHmcMessageHelper processHmcMessageHelper;
+    private final CcdCaseService ccdCaseService;
+
+    public void processEventMessageV2(HmcMessage hmcMessage)
+        throws CaseException, MessageProcessingException, InvalidMappingException {
+
+        Long caseId = hmcMessage.getCaseId();
+        String hearingId = hmcMessage.getHearingId();
+
+        HearingGetResponse hearingResponse = hmcHearingApiService.getHearingRequest(hearingId);
+
+        HmcStatus hmcMessageStatus = hmcMessage.getHearingUpdate().getHmcStatus();
+
+        if (processHmcMessageHelper.stateNotHandled(hmcMessageStatus, hearingResponse)) {
+            log.info("CCD state has not been updated using V2 for the Hearing ID {} and Case ID {}",
+                     hearingId, caseId
+            );
+            return;
+        }
+
+        log.info("Processing message using V2 for HMC status {} with cancellation reasons {} for the Hearing ID {} and Case ID"
+                     + " {}",
+                 hmcMessageStatus, hearingResponse.getRequestDetails().getCancellationReasonCodes(),
+                 hearingId, caseId
+        );
+
+        Function<SscsCaseDetails, DynamicEventUpdateResult> mutator = sscsCaseDetails -> {
+
+            SscsCaseData sscsCaseData = sscsCaseDetails.getData();
+
+            DwpState resolvedState = hearingUpdateService.resolveDwpState(hmcMessageStatus);
+            if (resolvedState != null) {
+                sscsCaseData.setDwpState(resolvedState);
+            }
+            if (processHmcMessageHelper.isHearingUpdated(hmcMessageStatus, hearingResponse)) {
+                try {
+                    hearingUpdateService.updateHearing(hearingResponse, sscsCaseData);
+                } catch (InvalidMappingException | MessageProcessingException e) {
+                    log.error("Error updating hearing using V2 for hearing ID {} and case ID {}", hearingId, caseId, e);
+                    throw new HearingUpdateException(e.getMessage(), e);
+                }
+            }
+
+            hearingUpdateService.setHearingStatus(hearingId, sscsCaseData, hmcMessageStatus);
+
+            hearingUpdateService.setWorkBasketFields(hearingId, sscsCaseData, hmcMessageStatus);
+
+            return resolveEventType(sscsCaseData, hmcMessageStatus, hearingResponse, hearingId);
+
+        };
+
+        SscsCaseDetails initialSscsCaseDetails = ccdCaseService.getCaseDetails(caseId);
+        IdamTokens idamTokens = idamService.getIdamTokens();
+        DynamicEventUpdateResult dynamicEventUpdateResult = mutator.apply(initialSscsCaseDetails);
```

Comment:
I suppose this is only invoked to get event type and whether to commit. While the actual case data update is done in the updateCaseV2DynamicEvent. Hence, if so than just invoking resolveEventType(initialSscsCaseDetails.getData(), hmcMessageStatus, hearingResponse, hearingId) should suffice.

Reply:
done